src/core/mesh/qgsmeshlayer.h

```diff
@@ -182,18 +183,23 @@ class CORE_EXPORT QgsMeshLayer : public QgsMapLayer
     const QgsMesh *nativeMesh() const SIP_SKIP;
 
     /**
-     * Returns triangular mesh (NULLPTR before rendering)
-     *
+     * Returns triangular mesh (NULLPTR before rendering). If the parameter triangleSize is provided, among the base triangular mesh
+     * and the simplified triangular meshes, the one returned is which has the average triangle size just greater than triangleSize.
+     * The size of a triangle is the maximum between the height and the width of the triangle bounding box
+     * \param triangleSize is the average size criteria in canvas map units
+     * \returns triangular mesh
+     * \note argument and simplified mesh was added in QGIS 3.14
      * \note Not available in Python bindings
      */
-    QgsTriangularMesh *triangularMesh() SIP_SKIP;
+    QgsTriangularMesh *triangularMesh( double triangleSize = 0 ) const SIP_SKIP;
 
     /**
-     * Returns triangular mesh (NULLPTR before rendering)
-     *
+     * Returns the base triangular mesh, create one if it doesn't exist
+     * \returns triangular mesh
      * \note Not available in Python bindings
+     * \since QGIS 3.14
      */
-    const QgsTriangularMesh *triangularMesh() const SIP_SKIP;
+    QgsTriangularMesh *baseTriangularMesh( const QgsCoordinateTransform &transform ) SIP_SKIP;
```

Comment:
- I am not convinced by argument `const QgsCoordinateTransform &transform` here. If the map layer knows its CRS and we know native mesh CRS, why we need the transform here? The function to create base triangular mesh from native mesh should be somewhere in utils, so anyone can create it from transform, but here in layer it should be done automatically? (if possible?)
- To keep the API backward compatible , I would keep the function triangularMesh() to return baseTriangular mesh as before and add TODO QGIS 4.x rename to baseTriangularMesh
- I would create function to simplified meshes with the average triangle parameter.

Reply:
This function returns the baseTriangularMesh. If this base triangular mesh is not created OR if the transform used is not the same, the function rebuilds the triangular mesh. It is used by other part of QGIS (analysis and 3D), so the transform could be different than the transform needed for map, for example in analysis which uses layer crs for source and destination.
I think it is a improvement of the old implementation which was not very clean to obtain the triangularMesh for other part than map renderer. But, I agree it still not very clean...
Maybe we can add an external function that create a triangular mesh from native mesh and QgsCoordinateTransform. The advantage will be that we could separate rendering map triangular mesh from other part. The disadvantage is that could lead to a lot of more memory consuming because we will have several triangular mesh in memory.

I think the API backward is kept, because we still have the method `triangularMesh() ` that provides the base mesh if no parameter is provided.

For the last point, do you mean that for simplified meshes creation, we continue to create simplified mesh until the average triangle size of the last simplified meshes is greater than the parameter ?

Reply:
can you please rename `baseTriangularMesh` to `updateTriangularMesh`, remove SIP_SKIP and to return `void` ?

so the worflow is to call `updateTriangularMesh` and that get triangular mesh by `triangularMesh` as before. That way we do not break API and it will be more readable

Reply:
I agree, your workflow is better and cleaner. I modify the code like you say